src/Connection/CountLimitingConnectionPool.php

```diff
@@ -0,0 +1,306 @@
+                }
+                return;
+            }
+
+            $this->openConnectionCount++;
+
+            if ($isHttps) {
+                $this->waitForPriorConnection[$uri] = \in_array('2', $connection->getProtocolVersions(), true);
+            }
+
+            $connection->onClose(function () use ($uri, $connectionId): void {
+                $this->openConnectionCount--;
+                $this->dropConnection($uri, $connectionId);
+            });
+        });
+
+        try {
+            $connection = yield Promise\first([$connectionPromise, $deferredPromise]);
+        } catch (MultiReasonException $exception) {
+            [$exception] = $exception->getReasons(); // The first reason is why the connection failed.
+            throw $exception;
+        }
+
+        $deferred = null; // Null reference so connection promise handler does not double-resolve the Deferred.
+        $this->removeWaiting($uri, $deferredId); // Deferred no longer needed for this request.
+
+        \assert($connection instanceof Connection);
+
+        $stream = yield $this->getStreamFrom($connection, $request);
+
+        if ($stream === null) {
+            // Reused connection did not have an available stream for the given request.
+            $connection = yield $connectionPromise; // Wait for new connection request instead.
+
+            $stream = yield $this->getStreamFrom($connection, $request);
+
+            // New connection should always return a stream for the request.
+            \assert($stream instanceof Stream);
+        }
+
+        return [$connection, $stream];
+    }
+
+    private function getStreamFrom(Connection $connection, Request $request): Promise
+    {
+        if (!\array_intersect($request->getProtocolVersions(), $connection->getProtocolVersions())) {
+            return new Success; // Connection does not support any of the requested protocol versions.
+        }
+
+        return $connection->getStream($request);
+    }
+
+    private function shouldMakeNewConnection(string $uri): bool
+    {
+        return \count($this->connections[$uri] ?? []) < $this->connectionLimit;
+    }
+
+    private function release(Connection $connection, string $uri): void
+    {
+        if (empty($this->waiting[$uri])) {
+            return;
+        }
+
+        $deferred = \array_shift($this->waiting[$uri]);
+        $deferred->resolve($connection);
+    }
+
+    private function removeWaiting(string $uri, int $deferredId): void
+    {
+        unset($this->waiting[$uri][$deferredId]);
+        if (empty($this->waiting[$uri])) {
+            unset($this->waiting[$uri]);
+        }
+    }
+
+    private function dropConnection(string $uri, int $connectionId): void
+    {
+        unset($this->connections[$uri][$connectionId]);
+
+        if (empty($this->connections[$uri])) {
+            unset($this->connections[$uri], $this->waitForPriorConnection[$uri]);
+        }
```

Comment:
If we drop a connection, shouldn't we do the same thing as `release` then?

Reply:
Not exactly, as that's suppose to be done when a connection is available for another request. I renamed the method accordingly.